File: src/main/java/com/pixelreel/service/provider/OmdbMovieSearchProvider.java
package com.pixelreel.service.provider;

import com.pixelreel.config.OmdbProperties;
import com.pixelreel.dto.external.ExternalMovieSearchResult;
import com.pixelreel.dto.external.MovieRecordSuggestion;
import com.pixelreel.dto.external.ProviderSearchResult;
import com.pixelreel.dto.omdb.OmdbSearchItem;
import com.pixelreel.dto.omdb.OmdbSearchResponse;
import com.pixelreel.entity.Movie;
import com.pixelreel.enums.RecordStatus;
import com.pixelreel.service.MovieService;
import com.pixelreel.service.search.MovieSearchProvider;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.client.RestClient;

@Service
@RequiredArgsConstructor
public class OmdbMovieSearchProvider implements MovieSearchProvider {
  private final MovieService movieService;

  @Qualifier("omdbRestClient")
  private final RestClient omdbRestClient;

  private final OmdbProperties omdbProperties;

  @Override
  public String id() {
    return "omdb";
  }

  @Override
  public ProviderSearchResult<ExternalMovieSearchResult> search(String query, int page) {
    ProviderSearchResult<ExternalMovieSearchResult> result = new ProviderSearchResult<>();
    result.setProvider(id());

    if (!omdbProperties.enabled()) {
      result.setEnabled(false);
      result.setMessage("OMDb 未启用");
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    if (!StringUtils.hasText(query)) {
      throw new IllegalArgumentException("query must not be blank");
    }

    int normalizedPage = Math.max(page, 1);

    OmdbSearchResponse response = omdbRestClient
        .get()
        .uri(uriBuilder -> uriBuilder
            .queryParam("apikey", omdbProperties.apiKey())
            .queryParam("s", query)
            .queryParam("page", normalizedPage)
            .queryParam("type", "movie")
            .build())
        .retrieve()
        .body(OmdbSearchResponse.class);

    if (response != null && "False".equalsIgnoreCase(response.getResponse())) {
      result.setEnabled(true);
      result.setMessage(response.getError());
      result.setPage(normalizedPage);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    List<OmdbSearchItem> items = response == null ? Collections.emptyList() : response.getSearch();
    Map<String, Movie> existingByImdbId = findExistingByImdbId(items);

    List<ExternalMovieSearchResult> results = new ArrayList<>();
    for (OmdbSearchItem item : items) {
      ExternalMovieSearchResult mapped = new ExternalMovieSearchResult();
      mapped.setProvider(id());
      mapped.setImdbId(item.getImdbId());
      mapped.setTitle(item.getTitle());
      mapped.setReleaseDate(item.getYear());
      mapped.setPosterUrl(normalizePoster(item.getPoster()));

      Movie existing = item.getImdbId() == null ? null : existingByImdbId.get(item.getImdbId());
      mapped.setAlreadyAdded(existing != null);
      mapped.setExistingRecordId(existing == null ? null : existing.getId());
      mapped.setSuggestedRecord(buildSuggestion(mapped));
      results.add(mapped);
    }

    int total = parseTotal(response == null ? null : response.getTotalResults());
    int totalPages = total == 0 ? 0 : (int) Math.ceil(total / 10.0);

    result.setEnabled(true);
    result.setPage(normalizedPage);
    result.setTotalPages(totalPages);
    result.setTotalResults(total);
    result.setResults(results);
    return result;
  }

  private Map<String, Movie> findExistingByImdbId(List<OmdbSearchItem> items) {
    List<String> ids = items.stream()
        .map(OmdbSearchItem::getImdbId)
        .filter(StringUtils::hasText)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return movieService.lambdaQuery()
        .in(Movie::getImdbId, ids)
        .list()
        .stream()
        .collect(Collectors.toMap(Movie::getImdbId, Function.identity(), (a, b) -> a));
  }

  private MovieRecordSuggestion buildSuggestion(ExternalMovieSearchResult mapped) {
    MovieRecordSuggestion suggestion = new MovieRecordSuggestion();
    suggestion.setImdbId(mapped.getImdbId());
    suggestion.setTitle(mapped.getTitle());
    suggestion.setPosterUrl(mapped.getPosterUrl());
    suggestion.setStatus(RecordStatus.WANT);
    suggestion.setRating(null);
    suggestion.setShortReview("");
    return suggestion;
  }

  private int parseTotal(String totalResults) {
    try {
      return totalResults == null ? 0 : Integer.parseInt(totalResults);
    } catch (NumberFormatException ex) {
      return 0;
    }
  }

  private String normalizePoster(String poster) {
    if (!StringUtils.hasText(poster) || "N/A".equalsIgnoreCase(poster)) {
      return null;
    }
    return poster;
  }
}

File: express-backend/src/services/provider/OmdbMovieSearchProvider.ts
import axios from 'axios';
import { config } from '../../config';
import { getDb } from '../../config/db';
import { MovieSearchProvider } from '../provider/MovieSearchProvider';
import {
  ExternalMovieSearchResult,
  ProviderSearchResult,
  MovieRecordSuggestion,
} from '../../dto/external-search';
import { RecordStatus } from '../../enums/RecordStatus';

function containsChinese(str: string): boolean {
  return /[一-鿿]/.test(str);
}

async function tryOmdbWithEnglishTitles(query: string, page: number): Promise<any | null> {
  try {
    const tmdbRes = await axios.get(`${config.tmdb.baseUrl}/search/movie`, {
      params: { query, page: 1 },
      headers: { Authorization: `Bearer ${config.tmdb.apiKey}` },
    });
    const results = tmdbRes.data?.results ?? [];
    // 按 popularity 降序排列，优先选最热门的结果
    const sorted = [...results].sort((a: any, b: any) => (b.vote_count ?? 0) - (a.vote_count ?? 0));
    // 收集所有英文候选标题（去重）
    const candidates: string[] = [];
    for (const item of sorted) {
      if (item.title && /[a-zA-Z]/.test(item.title) && !containsChinese(item.title)) {
        candidates.push(item.title);
      }
      if (item.original_title && /[a-zA-Z]/.test(item.original_title) && !containsChinese(item.original_title)
        && !candidates.includes(item.original_title)) {
        candidates.push(item.original_title);
      }
    }
    // 逐个尝试 OMDb 搜索，返回第一个有结果的
    for (const title of candidates) {
      const omdbRes = await axios.get(config.omdb.baseUrl, {
        params: { apikey: config.omdb.apiKey, s: title, page, type: 'movie' },
      });
      if (omdbRes.data?.Response === 'True') {
        return omdbRes.data;
      }
    }
    return null;
  } catch {
    return null;
  }
}

// OMDb 影视搜索 Provider，与 Java 端 OmdbMovieSearchProvider 完全对齐
export class OmdbMovieSearchProvider implements MovieSearchProvider {
  id(): string {
    return 'omdb';
  }

  async search(query: string, page: number): Promise<ProviderSearchResult<ExternalMovieSearchResult>> {
    const result: ProviderSearchResult<ExternalMovieSearchResult> = {
      provider: this.id(),
      enabled: true,
      message: '',
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };

    if (!config.omdb.apiKey) {
      result.enabled = false;
      result.message = 'OMDb 未启用';
      return result;
    }

    if (!query) throw new Error('query must not be blank');
    const normalizedPage = Math.max(page, 1);

    let response = await axios.get(config.omdb.baseUrl, {
      params: { apikey: config.omdb.apiKey, s: query, page: normalizedPage, type: 'movie' },
    });

    // 中文关键词或 OMDb 返回 Too many results 时，通过 TMDB 获取英文标题逐个重试
    if (containsChinese(query) || response.data?.Error === 'Too many results.') {
      const fallback = await tryOmdbWithEnglishTitles(query, normalizedPage);
      if (fallback) {
        response = { ...response, data: fallback };
      }
    }

    if (response.data?.Response === 'False') {
      result.message = response.data.Error ?? 'OMDb 搜索无结果';
      return result;
    }

    const items = response.data?.Search ?? [];
    const imdbIds = items.map((i: any) => i.imdbID).filter(Boolean);
    const existingMap = imdbIds.length > 0
      ? await this.findExistingByImdbId(imdbIds)
      : new Map<string, any>();

    const results: ExternalMovieSearchResult[] = items.map((item: any) => {
      const existing = item.imdbID ? existingMap.get(item.imdbID) ?? null : null;
      const poster = !item.Poster || item.Poster === 'N/A' ? null : item.Poster;
      const mapped: ExternalMovieSearchResult = {
        provider: this.id(),
        tmdbId: null,
        imdbId: item.imdbID ?? null,
        doubanId: null,
        traktId: null,
        title: item.Title ?? '',
        posterUrl: poster,
        releaseDate: item.Year ?? null,
        overview: null,
        alreadyAdded: existing !== null,
        existingRecordId: existing?.id != null ? Number(existing.id) : null,
        suggestedRecord: null,
      };
      mapped.suggestedRecord = this.buildSuggestion(mapped);
      return mapped;
    });

    const total = parseInt(response.data?.totalResults ?? '0', 10) || 0;
    const totalPages = total === 0 ? 0 : Math.ceil(total / 10);

    result.page = normalizedPage;
    result.totalPages = totalPages;
    result.totalResults = total;
    result.results = results;
    return result;
  }

  private async findExistingByImdbId(ids: string[]): Promise<Map<string, any>> {
    const movies = await getDb().movie.findMany({ where: { imdbId: { in: ids } } });
    return new Map(movies.map((m) => [m.imdbId!, m]));
  }

  private buildSuggestion(mapped: ExternalMovieSearchResult): MovieRecordSuggestion {
    return {
      tmdbId: mapped.tmdbId,
      imdbId: mapped.imdbId,
      doubanId: mapped.doubanId,
      traktId: mapped.traktId,
      title: mapped.title,
      posterUrl: mapped.posterUrl,
      status: RecordStatus.WANT,
      rating: null,
      shortReview: '',
    };
  }
}